### PR TITLE
Refactor and implement views

### LINKED ISSUE #23: Add UI widget for showing "history"
Collecting Audit for our record is great, but it needs to be displayed to a user or administrator at some point. We are willing to add a special ATK UI View to display history of changes for a specified model as well as give ability to add manual comment by the user (optional).

TODO - create a "History" view similar to this:
image

Differences:
 - most recent changes should appear on the top
 - form to leave comment should also be on the top (use inline form)
 - in our case modifications would usually affect multiple fields, so one line can contain multiple field value changes.
 - use css labels with field name and value. 
 - must interpret "json" data to decode fields old/new values but also display header with date/user info.
 - developer can set a different template with different tags

I recommend that Lister is used, however the render loop would include a rendering of the {Changes} field.

### PR DESCRIPTION
**Pull request by DarkSide666**
screenshot

Demo in /demos folder.

fix #23

### CODE REVIEW

**Comment by romaninsh** [comment]
Love it!

File: demos/index.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    return $c2->jsReload(['model_id'=>$id]);
+});
+
+
+
+// right side Audit History
+
+// create model for form
+$m2 = clone $m;
+if (isset($_GET['model_id'])) {
```
Comment (romaninsh) [inline demos/index.php]: if($id = $app->stickyGet('model_id')) {

..

File: src/Controller.php [hunk 2]
```diff
@@ -49,85 +87,118 @@ public function setUp(\atk4\data\Model $m)
...
         // save the initial action
         $a->save();
 
+        // memorize start time
         $a->start_mt = (float)microtime();
-        $m->audit_log = $a;
 
+        //Imants: deprecated - use $m->audit_log_controller->audit_log_stack[0] instead
+        // or $m->audit_log_controller->custom_action and custom_fields properties in your beforeSave hook
+        //$m->audit_log = $a;
```
Comment (romaninsh) [inline src/Controller.php]: why not keep a reference to `audit_log` ?

what if there are multiple audit logs?

File: template/audit-lister.html [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                <div class="date">{$ts}</div>
+            </div>
+            <div class="meta">
+                {$descr}
+            </div>
+            <div class="extra text">
+                {changes}
+                    {change}
+
+                    <!-- option 1 - show as segment with field name in header and labels inside -->
```
Comment (romaninsh) [inline template/audit-lister.html]: remove comments from {change}. Template is dumb and will add LOTS of those comments into output. 

Placing them between {changes} and {change} is best.

File: template/audit-history.pug [hunk 4]
```diff
@@ -0,0 +1,2 @@
+| <{_element}div{/} id="{$_id}" class="{_ui}ui{/} {$class} {_class}{/}" style="{$style}" {$attributes}>{$Content}
```
Comment (romaninsh) [inline template/audit-history.pug]: why first char is pipe? See: https://github.com/atk4/ui/blob/develop/template/semantic-ui/element.pug

File: template/audit-lister.pug [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      .summary
+        .user {$user}
+        |                 {$action}
+        .date {$ts}
+      .meta
+        | {$descr}
+      .extra.text
+        | {changes}
+        | {change}
+        // option 1 - show as segment with field name in header and labels inside
```
Comment (romaninsh) [inline template/audit-lister.pug]: remove comments from {change}. Template is dumb and will add LOTS of those comments into output.

Placing them between {changes} and {change} is best.

File: tests/BasicTest.php [hunk 6]
```diff
@@ -1,15 +0,0 @@
...
-namespace atk4\ui\tests;
-
-class BasicTest extends \atk4\core\PHPUnit_AgileTestCase
-{
-    /**
-     * Test constructor.
-     */
-    public function testTesting()
-    {
-        $this->assertEquals('foo', 'foo');
```
Comment (romaninsh) [inline tests/BasicTest.php]: ah, that was a good test 😆

File: tests/CustomTest.php [hunk 7]
```diff
@@ -19,14 +22,19 @@ function init()
 
         $this->addHook('beforeSave', function($m) {
             if ($m->isDirty('gender')) {
-                $m->audit_log['action'] = 'genderbending';
+                //$m->audit_log['action'] = 'genderbending'; // deprecated usage
+                //$m->audit_log_controller->audit_log_stack[0]['action'] = 'genderbending';
+                $m->audit_log_controller->custom_action = 'genderbending';
```
Comment (romaninsh) [inline tests/CustomTest.php]: can this be done through a method call? 

$m->addHistoryComment("genderbending");

File: demos/index.php (hunk 1 above)
Comment (DarkSide666) [reply demos/index.php]: addressed

File: template/audit-history.pug (hunk 4 above)
Comment (DarkSide666) [reply template/audit-history.pug]: fixed

File: template/audit-lister.html (hunk 3 above)
Comment (DarkSide666) [reply template/audit-lister.html]: fixed

File: template/audit-lister.pug (hunk 5 above)
Comment (DarkSide666) [reply template/audit-lister.pug]: fixed

File: tests/BasicTest.php (hunk 6 above)
Comment (DarkSide666) [reply tests/BasicTest.php]: yeah - 100% coverage 👯‍♂️

File: tests/CustomTest.php (hunk 7 above)
Comment (DarkSide666) [reply tests/CustomTest.php]: well - if that's in beforeSave hook, then it already have done audit->push() and set action='update'. so to change that at this point you have only one option - set `controller->custom_action` and/or `controller->custom_fields` properties which are respected in audit->pull().

I intentionally didn't want developer to mess directly with audit data model. He should use audit controller for that and audit controller should take care of audit data model.

File: src/Controller.php (hunk 2 above)
Comment (DarkSide666) [reply src/Controller.php]: I intentionally didn't want developer to mess directly with audit data model. He should use audit controller for that and audit controller should take care of audit data model.
I think that's better work-flow.

**Comment by DarkSide666** [comment]
@romans you can review https://github.com/atk4/audit/pull/24 again
• addressed your comments
• refactored migrator and demo now use Admin layout
• still don't see why we would need $model->audit_log if we have $model->auditController (see comment)
• fixed diff value of log record when deleting record from model
• renamed audit_log_controller to auditController
• added more tests for auditing of nested actions - see MultiModelTests
• not fixed (can't be fixed) - audit still will create empty log record even if there was no changes in data model. See https://github.com/atk4/audit/issues/17 (edited) 

New look:

**Pull request merged**